### ISSUE TITLE
unable to retrieve the complete list of server APIs: karpenter.sh/v1alpha5

### ISSUE THREAD

**Issue by akunduru9** [aws/karpenter-provider-aws#5425]
panic: failed to setup machine provider id indexer: failed to get API group resources: unable to retrieve the complete list of server APIs: karpenter.sh/v1alpha5: the server could not find the requested resource

goroutine 1 [running]:
github.com/samber/lo.must({0x25f73e0, 0xc0005f9800}, {0xc000b0f730, 0x1, 0x1})
        github.com/samber/lo@v1.38.1/errors.go:51 +0x1c5
github.com/samber/lo.Must0(...)
        github.com/samber/lo@v1.38.1/errors.go:72
github.com/aws/karpenter-core/pkg/operator.NewOperator()
        github.com/aws/karpenter-core@v0.32.3/pkg/operator/operator.go:182 +0x16ce
main.main()
        github.com/aws/karpenter/cmd/controller/main.go:33 +0x25

I am using latest tag  of karpenter helm charts and EKS version we have is 1.28

**Comment by jonathan-innis** [comment]
Which version of Karpenter are you running? You typically see this error when you don't have the Machine CRD installed, which means that the controller can't register the indexer that it needs to start running.

**Comment by akunduru9** [comment]
public.ecr.aws/karpenter/controller:v0.32.3@sha256:afa0d0fd5ac375859dc3d239ec992f197cdf01f6c8e3413e3845a43c2434621e is the version being used

**Comment by jonathan-innis** [comment]
You need the Machine CRD installed when you are running v0.32.x. This should already be included in the chart. How are you installing Karpenter?

**Comment by akunduru9** [comment]
"I am generating a manifest.yaml file using the update-manifest.sh script and applying it through GitOps. here is code snippet .#!/bin/bash

SCRIPTPATH="$( cd -- "$(dirname "$0")" >/dev/null 2>&1 ; pwd -P )" # https://stackoverflow.com/a/4774063
MANIFEST=${SCRIPTPATH}/manifest.yaml
CRDMANIFEST=${SCRIPTPATH}/crdmanifest.yaml

while [[ $# -gt 0 ]] && [[ $? -eq 0 ]]; do
  case "$1" in
    --version)
      VERSION="$2"
      shift
      shift
      ;;
    *)
      2>&1 echo "Unknown argument(s): $@"
      exit 1
      ;;
  esac
done

VALID_ARGUMENTS=1

if [[ -z "${VERSION}" ]]; then
  if ! command -v yq &> /dev/null; then
    2>&1 echo "--version is required when 'yq' is not available."
    VALID_ARGUMENTS=0
  else
    VERSION=$(curl --silent --fail --show-error --location --header "Accept: application/vnd.github+json" https://api.github.com/repos/aws/karpenter-provider-aws/releases | yq eval '.[] | select(.tag_name) | .tag_name' | head -1)
    if [[ -z "${VERSION}" ]]; then
      2>&1 echo "Error autodetecting release."
      VALID_ARGUMENTS=0
    fi
  fi
fi

if ! command -v helm &> /dev/null; then
  2>&1 echo "'helm' is required to render the manifest."
  VALID_ARGUMENTS=0
fi

if [[ "${VALID_ARGUMENTS}" != 1 ]]; then
  exit 1
fi

set -o errexit

echo "Updating manifest at ${MANIFEST} to ${VERSION}..."
curl -LOk "https://github.com/aws/karpenter-provider-aws/archive/refs/tags/${VERSION}.tar.gz"
tar -zxvf ${VERSION}.tar.gz -C ${SCRIPTPATH}

helm template karpenter "${SCRIPTPATH}/karpenter-provider-aws-${VERSION#v}/charts/karpenter" --namespace karpenter --values "${SCRIPTPATH}/karpenter-provider-aws-${VERSION#v}/charts/karpenter/values.yaml" >| "${MANIFEST}"

helm template karpenter "${SCRIPTPATH}/karpenter-provider-aws-${VERSION#v}/charts/karpenter-crd" --namespace karpenter  >| "${CRDMANIFEST}"

rm ${SCRIPTPATH}/${VERSION}.tar.gz
yes | rm -r "${SCRIPTPATH}/karpenter-provider-aws-${VERSION#v}"

**Comment by jonathan-innis** [comment]
It's hard for me to parse through the script to see what it's doing. Are you installing all of the CRDs that are part of the chart in the `crds/` directory? There's a callout for installing all of the CRDs for the chart here in the upgrade guide: https://karpenter.sh/v0.32/upgrading/upgrade-guide/#:~:text=In%20general%2C%20you%20can%20reapply%20the%20CRDs%20in%20the%20crds%20directory%20of%20the%20Karpenter%20helm%20chart

**Comment by akunduru9** [comment]
ya I am installing crds using this command helm template karpenter "${SCRIPTPATH}/karpenter-provider-aws-${VERSION#v}/charts/karpenter-crd" --namespace karpenter  >| "${CRDMANIFEST}"

**Comment by jonathan-innis** [comment]
I think the difference here is that you need it to install the CRDs based on the version that you are on. If you are installing from `main` for a version that isn't latest, this is going to break. Try https://github.com/aws/karpenter-provider-aws/tree/v0.32.4/charts/karpenter-crd/templates

**Comment by akunduru9** [comment]
sure do we also need to install chart we had here https://github.com/aws/karpenter-provider-aws/blob/v0.32.4/charts/karpenter-crd/Chart.yaml

**Comment by akunduru9** [comment]
used this command to execute helm template karpenter "${SCRIPTPATH}/karpenter-provider-aws-0.32.4/charts/karpenter-crd/templates" --namespace karpenter  >| "${CRDMANIFEST}"  it errored out saying Error: Chart.yaml file is missing

**Comment by jonathan-innis** [comment]
I think based on your script, you need to move up one directory so the `Chart.yaml` file is at the top of the path. The main point is that you need to point to the versioned thing, not just the latest thing. The version you are installing the CRDs from should match the version of the image you are installing.

**Comment by akunduru9** [comment]
I was able to manage and install versioned crds. this time i am seeing a different error any idea on that once here is error webhook.ValidationWebhook","message":"Reconcile error","commit":"3a61217","knative.dev/traceid":"454a0c6a-e510-4d5f-a0df-3fda8b797e9b","knative.dev/key":"validation.webhook.karpenter.sh","duration":"55.054059ms","error":"failed to update webhook: Operation cannot be fulfilled on validatingwebhookconfigurations.admissionregistration.k8s.io \"validation.webhook.karpenter.sh\": the object has been modified; please apply your changes to the latest version and try again

**Comment by jonathan-innis** [comment]
These are both expected. The first one is a transient error which is discussed here: #2902. The other one is just telling you that a Karpenter deployment can't run on capacity that is launched by Karpenter. This is to just keep you from having a cyclical dependency where you have Karpenter deployed to infra, but then the infra that Karpenter is running on goes bad and Karpenter can't recover itself because it manages the infra that it's running on.

If you are seeing that error continually, that means that you don't have enough static capacity (we recommend using MNG or Fargate) to run the Karpenter controller.

**Comment by akunduru9** [comment]
is there any documentation for using fargate to run karpenter.

**Comment by jonathan-innis** [comment]
Are you using MNG or Fargate? If MNG, how many nodes are provisioned as part of your MNG?

**Comment by akunduru9** [comment]
we are using managed node groups and currently we have two node groups provisioned

**Comment by akunduru9** [comment]
I was able to fix all the errors mentioned above, and the pods are running fine. However, when I tried to test Karpenter by deploying an app, it did not scale up. Below are the logs that I observed.

 
kubectl get events -n karpenter
LAST SEEN   TYPE      REASON              OBJECT                                            MESSAGE
20m         Warning   FailedCreate        replicaset/karpenter-test-deployment-5696f7bf6c   Error creating: Internal error occurred: failed calling webhook "consul-consul-connect-injector.consul.hashicorp.com": failed to call webhook: Post "https://consul-consul-connect-injector.consul.svc:443/mutate?timeout=10s": no endpoints available for service "consul-consul-connect-injector"
8m12s       Warning   FailedCreate        replicaset/karpenter-test-deployment-5696f7bf6c   Error creating: Internal error occurred: failed calling webhook "consul-consul-connect-injector.consul.hashicorp.com": failed to call webhook: Post "https://consul-consul-connect-injector.consul.svc:443/mutate?timeout=10s": no endpoints available for service "consul-consul-connect-injector"
9m34s       Normal    ScalingReplicaSet   deployment/karpenter-test-deployment              Scaled up replica set karpenter-test-deployment-5696f7bf6c to 6
9s          Warning   FailedCreate        replicaset/test-deployment-77b66f4c9f             Error creating: Internal error occurred: failed calling webhook "consul-consul-connect-injector.consul.hashicorp.com": failed to call webhook: Post "https://consul-consul-connect-injector.consul.svc:443/mutate?timeout=10s": no endpoints available for service "consul-consul-connect-injector"
5m37s       Normal    ScalingReplicaSet   deployment/test-deployment                        Scaled up replica set test-deployment-77b66f4c9f to 2
